ci/codecov.sh

```diff
@@ -23,51 +23,69 @@ set -ex
     if [ -z "$GCOV" ]; then
-        ver=7 # default
-        if [ "${B2_TOOLSET%%-*}" == "gcc" ]; then
-            if [[ "$B2_TOOLSET" =~ gcc- ]]; then
-                ver="${B2_TOOLSET##*gcc-}"
-            elif [[ "$B2_COMPILER" =~ gcc- ]]; then
-                ver="${B2_COMPILER##*gcc-}"
+        # enforce.sh ensures B2_COMPILER
+        COMPILER=${B2_COMPILER}
+        COMPILER_FAMILY=$(echo ${COMPILER} | cut -d'-' -f1)
+        COMPILER_EDITION=$(echo ${COMPILER} | cut -d'-' -f2)
+        # clang is still not working - need a llvm-cov wrapper and fixup pathing
+        if [ "${COMPILER_FAMILY}" == "clang" ]; then
+            echo "clang is not supported for coverage yet"
+            exit 1
+            # find /usr/lib -name 'llvm-cov*' -print
+            # LLVM_COV_PATH=$(find /usr/lib -type f -name llvm-cov -exec dirname {} \;)
+            # export PATH=${LLVM_COV_PATH}:${PATH}
+            # GCOV=llvm-cov
+        elif [ "${COMPILER_FAMILY}" == "gcc" ]; then
+            # cutting "gcc" above yields "gcc" for both values if no version present
+            if [ "${COMPILER_EDITION}" == "${COMPILER_FAMILY}" ]; then
+                GCOV=gcov
+            else
+                GCOV=gcov-${COMPILER_EDITION}
             fi
+        else
+            echo "Cannot determine GCOV."
+            exit 1
         fi
-        GCOV=gcov-${ver}
     fi
 
     # install the latest lcov we know works
     rm -rf /tmp/lcov
     cd /tmp
-    git clone --depth 1 -b v1.14 https://github.com/linux-test-project/lcov.git
+    git clone --depth 1 -b v1.15 https://github.com/linux-test-project/lcov.git
     export PATH=/tmp/lcov/bin:$PATH
     command -v lcov
     lcov --version
 
     # switch back to the original source code directory
     cd $BOOST_CI_SRC_FOLDER
-    : "${LCOV_BRANCH_COVERAGE:=1}" # Set default
+    : "${LCOV_BRANCH_COVERAGE:=1}" # On by default, job can override but advise against it
 
     # coverage files are in ../../b2 from this location
-    lcov --gcov-tool=$GCOV --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --base-directory "$BOOST_ROOT/libs/$SELF" --directory "$BOOST_ROOT" --capture --output-file all.info
-    # dump a summary on the console
-    lcov --list all.info
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --gcov-tool=${GCOV} --directory "${BOOST_ROOT}" --capture --output-file all.info
+    lcov --rc lcov_branch_coverage=${LCOV_BRANCH_COVERAGE} --list all.info  # mostly for debug but interesting to see everything captured
 
-    # all.info contains all the coverage info for all projects - limit to ours
-    # first we extract the interesting headers for our project then we use that list to extract the right things
+    # this is tricky; not all of the headers in the project go into a directory with the project
+    # name, so we have to itemize all of them and extract them, along with anything in the project
+    # directory including tests - after this we drop coverage on test files to finalize things
```

Comment:
To be exact: None of the headers in "the project" (naming confused me) are in the projects folder, or better: The coverage data is not for those headers but for the copies/symlinks.

Maybe:

```suggestion
    # all.info contains all the coverage info for all projects - limit to ours:
    # Our headers are copied/symlinked into the top-level include folder (in $BOOST_ROOT) and coverage data references those paths.
    # So gather a list of our header file names and extract the coverage data for those and everything in our project folder
```